### PR TITLE
[BUGFIX beta] restore didInitAttrs + didReceiveAttrs to POST_INIT

### PR DESCRIPTION
**Pull request by stefanpenner**
- [x] test POST_INIT
- [X] restore hooks to POST_INIT
- [x] tests for `didInitAttrs + didReceiveAttrs` during POST_INIT
- [x] cleanup
- [x] feedback / review / 🚲 🏚 
- [x] test in offending app, going to delegate this to -> @chadhietala  / @krisselden

### CODE REVIEW

File: packages/ember-glimmer/lib/syntax/curly-component.js
Comment (stefanpenner) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: such, WAT

Comment (rwjblue) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: WAT indeed, why are you doing this?

Comment (stefanpenner) [inline packages/ember-glimmer/lib/syntax/curly-component.js]: quick hack, turns out `attrs` is on `this`

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: what do you think if we change the test to...

```js
this.registerComponent('foo-bar', {
  template: '{{foo}}-{{bar}}',

  ComponentClass: Component.extend({
    didInitAttrs({ attrs }) {
      this.set('bar', attrs.foo + 1);
    },

    barDidChange: ...
  })
});

assert.equal(fooDidChangeCount, 0);

this.render('{{foo-bar foo=foo}}', { foo: 1 });

assert.equal(fooDidChangeCount, 0);
this.assertText('1-2');

...

this.runTask(() => set(this.context, 'foo', 2);

...

```

Benefits (imo):

1. avoids the saving off `this` init dance
2. uses the attrs arg
3. (subjective) more meaningful/realistic

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: ditto

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: Can we add a test ~ like this?

```js
// This is not fully supported, but some component unit tests might rely on it?
['@test did{Init,Receive}Attrs fires on init without rendering/appending to DOM']() {
  expectDeprecation(/didInitAttrs called/);

  let didInitAttrsFired = 0;
  let didReceiveAttrsFired = 0;

  let MyComponent = Component.extend({
    didInitAttrs() {
      didInitAttrsFired++;
    },
    didReceiveAttrs() {
      didReceiveAttrsFired++;
    }
  });

  this.component = Component.create();

  // assert  === 1 for both
}
```

Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: Just to keep track of the implicit requirements, to make things easier when revisiting in the future

**Review by chancancode — CHANGES_REQUESTED** [review]
👍

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (chancancode) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: To do it properly, we will need to `this.registerComponent('foo-bar', { ComponentClass: ... });` + this.owner._lookupFactory. I don't think we support creating components w/o the injections – but that likely doesn't matter for this test as those restrictions are about ensuring the renderer (etc) are present, and we are explicitly not rendering here. The whole scenario is pretty wat and we should deprecate component unit tests and remove this test ¯\_(ツ)_/¯

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (stefanpenner) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: I would prefer to avoid rendering entirely, the timings this is testing occurs well before rendering.

If there is a better file to put more focused tests, please direct me to it.

Comment (stefanpenner) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: I have removed rendering entirely from the mix.

**Comment by stefanpenner** [comment]
@chancancode tests changed.

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: Manually creating components like this is a special case scenario (also tested, but not in this file), we should continue to use `render` to test this in the actual way that we care about this problem.  

Specifically, the thing observed by your work app had nothing to do with manually calling `.create()` (AFAIK), it was specifically that during initial render we trigger events/observers/etc.  That is the thing to test here...

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (rwjblue) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: Can you tweak this to avoid the "inside baseball" of `finishChains`?  Would you be ok with something like:

`@test did{Init,Receive}Attrs fires after .init() but before events are triggered and observers are setup`

I fear using `finishChains` will lead to confusion about what this test is actually intended to test over time.

**Review by rwjblue — CHANGES_REQUESTED** [review]

File: packages/ember-glimmer/tests/integration/components/curly-components-test.js
Comment (stefanpenner) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: the test should be tested via manual creation, which file is better suited for it. I will move this test there.

Comment (stefanpenner) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: I can do both, but according to our own documenting this is public: http://emberjs.com/api/classes/Ember.Component.html#method_create

Comment (stefanpenner) [inline packages/ember-glimmer/tests/integration/components/curly-components-test.js]: @rwjblue I can add another test that tests rendering, but given that this "in the wild" i would prefer to also keep around the very focused test.

**Review by rwjblue — APPROVED** [review]

**Comment by stefanpenner** [comment]
got +1 from @chancancode in chat

**Comment by stefanpenner** [comment]
testing the cherry-pick to beta now

**Comment by stefanpenner** [comment]
https://github.com/emberjs/ember.js/commit/c5ac34b5d52cfaabb7a43214f4a9709a33ed90e1

waiting on: https://travis-ci.org/emberjs/ember.js/builds/166819104 for the beta build of this

**Pull request merged**
